### ISSUE TITLE
Non Session Manager with Client-Server application

### ISSUE THREAD

**Issue by 0mk** [original-male/non#193]
I have a client-server application written in qt/c++. I have a class in the client to handle non session management. It works well except for one issue. There is a button on the client gui for re-starting the application's server process. When I press the button, the client kills the server via QProcess::kill(). This should result in a QProcess::finished signal being received by the client. The problem is that this signal never happens. The server process is left in a zombie state (defunct) and QProcess::state() is set to QProcess::Running.  If I run the application normally (not under NSM), it works fine; the server dies properly and the client can then start a new one.   Any ideas on why running under NSM should alter the behaviour in this way? I can supply example code for the problem if you would like.

**Comment by 0mk** [comment]
Solved it myself. The issue was that nsmd sets its signal mask to block SIGCHLD and that signal mask gets inherited by the clients that it launches. The solution is to unblock SIGCHLD in my application so it receives notification when ITS children die.

**Comment by ViktorNova** [comment]
I am very interested in this project of yours (I have been attempting a web-based NSM frontend myself but haven't gotten far), do you have your code published anywhere?

**Comment by 0mk** [comment]
Hi Viktor,


I have attached a small file with sample source code which I hope may be 
of use.


Regards,  Martin

**Comment by 0mk** [comment]
Oops, That obviously doesn't work (attachments not supported). If you 
give me a normal email address, I'll send you the source code.

Regards,  Martin.

**Comment by nico202** [comment]
@0mk maybe you can upload it as a [gist](https://gist.github.com/), i was interested too

**Comment by 0mk** [comment]
Hi Nicolo,


I want to upload/link/email  a tar file (tar.bz2). Gist seems to only 
support directly entered text.

Regards, Martin

**Comment by 0mk** [comment]
OK, try the following link:-

https://github.com/0mk/Spoon-Knife/blob/master/NSMClient.tar.bz2


Regards,  Martin

**Comment by 0mk** [comment]
Did you get the files ok? Were they useful. Please let me know.

Regards,  Martin (0mk)

**Comment by ViktorNova** [comment]
@0mk, thank you for sharing!
I downloaded the file but am unable to extract it, tar -xf gives this error:
> bzip2: (stdin) is not a bzip2 file.
tar: Child returned status 2
tar: Error is not recoverable: exiting now

I tried renaming to .tar.gz and .tar and .zip, none of which worked with their respective unextracting tools.. Can you try re-compressing and sending again? Thank you!
-Viktor

**Comment by 0mk** [comment]
Hi Viktor


I downloaded it myself and successfully extracted it, both with

tar xvf NSMClient.tar.bz2 AND with

tar -xf NSMClient.tar.bz2,

so there is definitely nothing wrong with the file.

There is no need to rename the file or use another tool.

I am using Gnu tar version 1.27 on Ubuntu. Please make sure your version 
of tar is compatible.


Also, you could try a two-stage extraction as in:


bunzip2 NSMClient.tar.bz2

followed by

tar xvf NSMClient.tar


Please let me know how it goes.


Regards,  Martin

**Comment by ViktorNova** [comment]
Aha, I realized my problem, apparently downloading your file using wget corrupts it somehow.. (makes no sense?) I downloaded it through the browser and extracted it and compiled with no issues

The UI runs and displays a window with a red X button and what looks like a play button in the upper right. Beneath these buttons is the text 
````NsmClient: Constructor()````
The X button closes the program and the play button outputs ````Test button pressed```` below the existing text.. I can't make much sense of it at this point, but I assume it's early stages =)

Are you going to put this project on GitHub? I would love to help test and give feedback, this is a project that interests me very much! Although I'm not much help with C++, I'm ok at QT Creator and have an eye for design and would love to help any way I can.

**Comment by 0mk** [comment]
Viktor,


It's not meant to be a "real world" application in the sense of it being 
useful for anything very much :-D.

It's purpose is simply to demonstrate the code needed for making an NSM 
client application. If you run it under the NSM session manager, you 
will see there is more output in its console window as it executes the 
NSM protocol when run under NSM session manager. None of that happens 
when it is run stand-alone. I provided the code for you because you said 
you were trying to make an NSM application and so I thought the source 
code might be useful.

If you don't know C++, it won't be much use I'm afraid. However, that 
being said, it's good to hear that you are so keen to help out with open 
source projects in general. I commend you for that. You might wish to 
invest some time to learn C++. There's a lot to learn. I'm fairly adept 
at it but certainly no expert. You certainly don't need to know the 
entire language in every detail to start using it productively though. 
There are plenty of online tutorials for people of all levels of 
experience so be bold and give it a go!

If you learn enough, you should be able to understand how my application 
works and what it does in more detail. Good luck!

Martin.

**Comment by ViktorNova** [comment]
Martin, thank you for the explanation.
Somehow I misinterpreted your initial description to mean that you are working on an alternative front-end to the NSM server (instead of the included FLTK frontend) and I got excited, as I have been slowly working on just that, only in Python 
(I may scrap it and start over in Node.js) 

Thank  you for your encouragement. I will probably stick to Python and learning Node.js, as C++ will involve more time than I currently have to invest in coding.. =\ I don't know if you do any Python, but I have found these NSM helper modules that I think do a similar thing to what you've wrote, only for Python:
https://github.com/hilbrichtsoftware/pynsmclient
https://github.com/hilbrichtsoftware/pynsm2

Cheers!

-Viktor
